strawberryfields/io.py

```diff
@@ -244,6 +246,108 @@ def is_free_param(param):
     return prog
 
 
+def generate_code(prog, eng=None):
+    """Converts a Strawberry Fields program into valid Strawberry Fields code.
+
+    Args:
+        prog (Program): the Strawberry Fields program
+        eng (Engine): the Strawberryfields engine
+
+    Returns:
+        str: the Strawberry Fields code, for constructing the program, as a string
+    """
+    code_seq = ["import strawberryfields as sf", "from strawberryfields import ops\n"]
+
+    if prog.type == "tdm":
+        code_seq.append(f"prog = sf.TDMProgram(N={prog.N})")
+    else:
+        code_seq.append(f"prog = sf.Program({prog.num_subsystems})")
+
+    # check if an engine is supplied; if so, format and add backend/target
+    # along with backend options
+    if eng:
+        eng_type = eng.__class__.__name__
+        if eng_type == "RemoteEngine":
+            code_seq.append(f'eng = sf.RemoteEngine("{eng.target}")')
+        else:
+            if eng.backend_options:
+                formatting_str = (
+                    f'"{eng.backend_name}", backend_options='
+                    + f'{{"cutoff_dim": {eng.backend_options["cutoff_dim"]}}}'
+                )
+                code_seq.append(f"eng = sf.Engine({formatting_str})")
+            else:
+                code_seq.append(f'eng = sf.Engine("{eng.backend_name}")')
+
+    # check if program is of TDM type and format the context as appropriate
+    if prog.type == "tdm":
+        tdm_params = [f"[{_factor_out_pi(par)}]" for par in prog.tdm_params]
```

Comment:
How come pi is factored out here?

Reply:
Ah! I read the function below and now get it. Might make sense to add a one line comment here just for clarity